x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java

```diff
@@ -179,17 +204,23 @@ public void beforeBulk(long executionId, BulkRequest request) {
             public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
                 numIndexed.addAndGet(request.numberOfActions());
                 if (response.hasFailures()) {
-                    Map<String, String> failures = Arrays.stream(response.getItems())
+                    List<BulkItemResponse> failedItems = Arrays.stream(response.getItems())
                         .filter(BulkItemResponse::isFailed)
+                        .collect(Collectors.toList());
+                    numFailed.addAndGet(failedItems.size());
+
+                    Map<String, String> failures = failedItems.stream()
                         .collect(
                             Collectors.toMap(
                                 BulkItemResponse::getId,
                                 BulkItemResponse::getFailureMessage,
                                 (msg1, msg2) -> Objects.equals(msg1, msg2) ? msg1 : msg1 + "," + msg2
                             )
                         );
-                    numFailed.addAndGet(failures.size());
                     logger.error("Shard [{}] failed to populate rollup index. Failures: [{}]", indexShard.shardId(), failures);
+
+                    // cancel rollup task
+                    abort = true;
```

Comment:
I think we can just call `task.cancel(...)` here and maybe use the log message as reason.